### PR TITLE
Upgrade ansilexer color support

### LINKED ISSUE #148: ansilexer custom lexer/formatter colorizes pytest output poorly
Love the theme/plugin/css/ etc. Thanks.

However, I was trying to include some pytest output in a blog post and found the output to be badly colorized.

Text as rendered at the console by pytest:
image

and as rendered in the blog post by `include`:
image

output from  `pygmentize -v -x -l .\m.css\plugins\ansilexer.py:AnsiLexer -f .\m.css\plugins\ansilexer.py:HtmlAnsiFormatter .\pytest.txt`:

```
============================= test session starts =============================
platform win32 -- Python 3.8.1, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: D:\Sandbox\LibraryHippo
plugins: requests-mock-1.8.0
collected 2 items

tests\unit\test_wpl.py .<span class="g g-AnsiGreen">.
         [100%]</span>

============================== 2 passed<span class="g g-AnsiGreen"> in 1.46s ==============================</span>
```
As you can see, there's no directive to colorize
* the first line
* the first `.` on the test\unit\test_wpl.py line
* the first half of the summary line

(I'm actually quite surprised that the first `.` on the test\unit\test_wpl.py line appears green in the output!)

The fault appears to be in the lexer, as when I run
`pygmentize -x -l .\m.css\plugins\ansilexer.py:AnsiLexer -f raw  pytest.txt`, I see

```code
Token.Text	'============================= test session starts ============================='
Token.Text	'\nplatform win32 -- Python 3.8.1, pytest-5.4.2, py-1.8.1, pluggy-0.13.1\nrootdir: D:\\Sandbox\\LibraryHippo\nplugins: requests-mock-1.8.0\ncollected 2 items\n\ntests\\unit\\test_wpl.py '
Token.Text	'.'
Token.Generic.AnsiGreen	'.'
Token.Generic.AnsiGreen	'                                                [100%]'
Token.Text	'\n\n'
Token.Text	'============================== '
Token.Text	''
Token.Text	'2 passed'
Token.Generic.AnsiGreen	' in 1.46s'
Token.Generic.AnsiGreen	' =============================='
Token.Text	'\n'
```

[pytest.txt](https://github.com/mosra/m.css/files/4656422/pytest.txt)

I'm happy to investigate, although I've had problems running the m.css tests, possibly due to missing prereqs (and I'm running on Windows).

### PR DESCRIPTION
**Pull request by blairconrad**
Fixes #148.

### CODE REVIEW

**Comment by blairconrad** [comment]
Notes:

1. I haven't built the docs. I'm a bad person.
2. The ansilexer is essentially completely rewritten. Sorry about that.
3. I made some opinionated choices rather than bombarding you with questions. **I will make any alterations you like**. Do not hesitate to get me to change anything at all.
4. One of the perhaps controversial things I did was replace `console-colors.ansi` with a beefed up version. I did this in a separate commit, so it's easy to drop and restore it.

File: css/pygments-console.css [hunk 1]
```diff
@@ -1,4 +1,21 @@
...
+.m-console .g-AnsiBackgroundBrightRed { background-color: #c0392b } /* Generic.AnsiBackgroundBrightRed */
+.m-console .g-AnsiBackgroundBrightWhite { background-color: #ffffff } /* Generic.AnsiBackgroundBrightWhite */
+.m-console .g-AnsiBackgroundBrightYellow { background-color: #fdbc4b } /* Generic.AnsiBackgroundBrightYellow */
+.m-console .g-AnsiBackgroundCyan { background-color: #1abc9c } /* Generic.AnsiBackgroundCyan */
+.m-console .g-AnsiBackgroundDefault { background-color: #fcfcfc } /* Generic.AnsiBackgroundDefault */
+.m-console .g-AnsiBackgroundGreen { background-color: #11d116 } /* Generic.AnsiBackgroundGreen */
+.m-console .g-AnsiBackgroundMagenta { background-color: #9b59b6 } /* Generic.AnsiBackgroundMagenta */
+.m-console .g-AnsiBackgroundRed { background-color: #ed1515 } /* Generic.AnsiBackgroundRed */
+.m-console .g-AnsiBackgroundWhite { background-color: #fcfcfc } /* Generic.AnsiBackgroundWhite */
+.m-console .g-AnsiBackgroundYellow { background-color: #f67400 } /* Generic.AnsiBackgroundYellow */
```
Comment (blairconrad) [inline css/pygments-console.css:18]: I assumed you'd want these analogous background classes…

File: plugins/ansilexer.py [hunk 2]
```diff
@@ -28,83 +29,288 @@
...
+
+        if (self._bright
+            and not self._foreground.startswith('Bright')
+            and not self._foreground.startswith('ForegroundColor')):
+            token = 'Generic.AnsiBright' + self._foreground
         else:
+            token = 'Generic.Ansi' + self._foreground
+
+        if (self._background != 'Default'):
+            token += 'BackgroundColor' + self._background
```
Comment (blairconrad) [inline plugins/ansilexer.py:216]: The BackgroundColor section of the token name has a different convention than the foreground, as this prefix always appears, to distinguish from the foreground color. I assumed that the representation of the tokens wasn't a huge deal, so long as the formatter spits out the right thing…

File: plugins/ansilexer.py [hunk 3]
```diff
@@ -28,83 +29,288 @@
...
+        if i >= 0:
+            foreground_color = ansi_class[i + 15:]
+            ansi_class = ansi_class[:i]
+            html_styles.insert(0, 'color: #' + foreground_color)
+
+        if ansi_class != 'g-Ansi':
+            html_classes.insert(1, ansi_class)
+
+        result = ''
+        if len(html_classes) > 1: # we don't want to emit just g
```
Comment (blairconrad) [inline plugins/ansilexer.py:302]: Previously, you'd dropped the `g` when emitting an RGB-styled foreground color via the style attribute, so I assumed this was intentional and drop it whenever there's no class that contributes to the colour.

File: plugins/ansilexer.py [hunk 4]
```diff
@@ -28,83 +29,288 @@
...
+                html_classes.append('g-AnsiBackground' + background_color)
+
+        i = ansi_class.find('ForegroundColor')
+        if i >= 0:
+            foreground_color = ansi_class[i + 15:]
+            ansi_class = ansi_class[:i]
+            html_styles.insert(0, 'color: #' + foreground_color)
+
+        if ansi_class != 'g-Ansi':
+            html_classes.insert(1, ansi_class)
```
Comment (blairconrad) [inline plugins/ansilexer.py:299]: The `insert(1…` here is mostly to keep the classes in the original order so as not to perturb the unit tests too much. Also, I guess it's nice to have foreground before background…

File: plugins/ansilexer.py [hunk 5]
```diff
@@ -28,83 +29,288 @@
...
+            if result:
+                result += ' '
+            result += 'style="' + '; '.join(html_styles) + '"'
+
+        # Add ZWNJ at the start of each span because otherwise it's
+        # somehow impossible to wrap even with word-break: break-all.
+        # Not sure why (and not sure if this is the best solution), but
+        # had to ship a thing so there it is. Adding <wbr/> did not
+        # help.
+        result += '>&zwnj;'
```
Comment (blairconrad) [inline plugins/ansilexer.py:315]: I'll admit I didn't understand this stuff at all. You'd been inserting it only when emitting RGB-defined colors, and I... didn't really know what to do, so I guessed, and am emitting everywhere. Happy to change.

File: plugins/m/test/code/console-colors.ansi [hunk 6]
```diff
@@ -1,5 +1,99 @@
...
+  [4mset underscore[0m
+  [5mset blink[0m
+  [7mset reverse video[0m
+  [10mreset selected mapping, display control flag, and toggle[0m
+  [11mselect null mapping, set display control flag, reset[0m
+  [12mselect null mapping, set display control flag, set toggle[0m
+  [21mset underline[0m
+  [24munderline off[0m
+  [25mblink off[0m
+  [27mreverse video off[0m
```
Comment (blairconrad) [inline plugins/m/test/code/console-colors.ansi:99]: That's not easy to read. The relevant section of `page.html` looks like this

image

File: plugins/ansilexer.py [hunk 7]
```diff
@@ -28,83 +29,288 @@
...
-            token += 'Cyan'
-        elif color == '37':
-            token += 'White'
-        elif color == '39':
-            token += 'Default'
         else:
-            yield (match.start(), Text, text)
-            return
+            yield (match.start(), string_to_tokentype(token), text)
+        return
```
Comment (sizmailov) [inline plugins/ansilexer.py:222]: redundant return

Comment (blairconrad) [reply plugins/ansilexer.py]: How embarrassing. Amended and pushed. Original branch renamed to ansilexer-pre-amend for your comparing convenience.

**Comment by sizmailov** [comment]
You mentioned you run on windows. Is there a need in testing on linux? I can help with that.

**Comment by blairconrad** [comment]
Oh, that's kind, @sizmailov. I started trying to set up an environment on WSL and got sidetracked. (I'm secretly wondering whether a layer of build automation couldn't be applied to the repo, but that's another discussion). I think the only tests I'd've affected are for the code plugin, and I was able to confirm that they still pass. _If_ you wanted to run something, you could build the docs and make sure the Math and Code section renders…
I should really try to set myself up to build.

**Comment by sizmailov** [comment]
Apparently travis-ci integration is broken and github doesn't show builds. Here is [direct link](https://travis-ci.org/github/mosra/m.css/pull_requests). I'll look into it locally.

**Comment by blairconrad** [comment]
Ah! Thanks for the link. It looks like I broke some doxygen tests that involve ansi codes. Probably I just need to update the reference file. I can look at it later today.

**Comment by sizmailov** [comment]
Indeed, I've run tests locally and replaced files, which made almost all test pass (tests with math result to arbitrarlly formatted svg, so they never run successfully on my machine). I've submitted a PR to you PR :)

**Comment by blairconrad** [comment]
Oh, thanks, @sizmailov. In parallel, I'd fixed some of the tests and updated files. Your changes are much better. I am inclined to take them. Am struggling with how to merge in while preserving your credit!

**Comment by sizmailov** [comment]
Don't worry about credits, they are not worth wasted time.

**Comment by blairconrad** [comment]
https://travis-ci.org/github/mosra/m.css/builds/696083412

> Don't worry about credits, they are not worth wasted time.

I feel somewhat bad about this, but would also say exactly the same thing in your situation, so I've incorporated your changes, with thanks!
Unsurprisingly, we have success! https://travis-ci.org/github/mosra/m.css/builds/696083412

File: plugins/ansilexer.py (hunk 5 above)
Comment (mosra) [reply plugins/ansilexer.py:314]: Yeah, sorry. This was to make the wrapped terminal output work on this post: https://blog.magnum.graphics/backstage/multidimensional-strided-array-views/ -- without the `&zwnj;`  the browser didn't really want to wrap the output for some reason, putting scrollbars on the thing, and ugh.

I'm adding "hooks" for code block postprocessing so I guess the ZWNJ isn't really needed anymore and I'll do it on my side instead. Can you drop it, then? Thanks.

**Comment by mosra** [comment]
Wow, man. This is beyond great, I'm impressed.

I'll remove the ZWNJ, fix the conflict (caused by me updating license headers) and merge. Thanks a lot.

**Comment by mosra** [comment]
Merged as f3a166100128107527a0335c6912c4bc9754831e (I removed the ZWNJ hack before in d1837841fc4705f311a106c2efe553f811c46a9d, to have this patch clean). Thanks again!

**Comment by blairconrad** [comment]
Thanks for the kind words, @mosra, and the opportunity to contribute (for my own benefit, as I'm the one who was complaining). And thanks again @sizmailov. I'd've never've known I broke tests without you.

I should probably try to get a Linux VM or WSL or something running…

**Comment by mosra** [comment]
With this, m.css has ANSI color code parsing better than most CIs (neither Travis, CircleCI nor AppVeyor were able to properly decode everything I wanted them to), so thank *you* :)

Regarding testing -- I'm fine if you abuse the CI for that (I'm an extreme on the other side, maintaining a project that runs on Windows and booting Windows only once or twice a year, mostly just relying on the CIs for that).

Btw., one thing I noticed in the "ground truth" output:

```html
<span class="g g-AnsiDefault" style="background-color: #5f00d7">
```

(And then repeated several times for all the colors.) Isn't the `g g-AnsiDefault` redundant in this case? I didn't look close enough to see if this is really an explicitly specified default in the input (in which case I'd say it's okay), but if not then I think it isn't needed. Or is this a nested override of something defined above and thus it needs to be there?

**Comment by blairconrad** [comment]
You... may be right. The `g g-AnsiDefault` might be superfluous here. I can take a look tonight or tomorrow.

**Comment by mosra** [comment]
Thanks! It's not a critical issue, just that it generates output that's heavier than it could be.

**Pull request closed without merging**